Nearly complete rewrite to switch from pgEdge CLI to native packages.

bonesmoses opened the pull request:
The CLI tool will eventually be deprecated, and pgEdge now produces official OS distribution packages for RHEL and Debian family Linux systems. This patch is an effort to switch to native integration with these tools. This includes deferring to the postgres OS user when possible, and no longer using the remote Ansible user as the owner of database and other files.

Several new variables were introduced to account for default binary and data directories, as well as operating system users. Existing defaults have been demonstrated as functional on both RHEL 9 and Debian 12 cloud images.

Addresses JIRA Issue EE-22.

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Summary by CodeRabbit

## Release Notes

* **New Features**
  * HAProxy can now run directly on pgEdge nodes via a separate configurable port
  * Customizable pg_hba.conf rules for fine-grained access control
  * Dedicated backup user for targeted SSH-based backup access

* **Major Changes**
  * CLI component replaced with standard RPM/DEB package installation
  * Standard postgres OS user replaces local-user installation model
  * Enhanced backup infrastructure with configurable user and repository parameters
  * Improved cluster initialization with better Patroni integration

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

bonesmoses committed:
Nearly complete rewrite to switch from pgEdge CLI to native packages.

The CLI tool will eventually be deprecated, and pgEdge now produces
official OS distribution packages for RHEL and Debian family Linux
systems. This patch is an effort to switch to native integration with
these tools. This includes deferring to the postgres OS user when
possible, and no longer using the remote Ansible user as the owner of
database and other files.

Several new variables were introduced to account for default binary
and data directories, as well as operating system users. Existing
defaults have been demonstrated as functional on both RHEL 9 and Debian
12 cloud images.

Addresses JIRA Issue EE-22.

bonesmoses committed:
Fixed SSH service name on RHEL systems for service reload.

bonesmoses committed:
Fixed issue where proxy_port wasn't being properly configured.

bonesmoses committed:
Based on code review by Claude:

* Explicitly reload systemd-logind rather than ssh after RemoveIPC disabled
* Now wait up to 5 minutes to establish initial Patroni cluster

bonesmoses commented on README.md:
```diff
@@ -40,15 +40,21 @@ All of the roles are meant to operate in conjunction. They are simplified to red
 | pg_version | 17 | Postgres version to install. Default to the most recent version. |
 | spock_version | 5.0.0 | Version of the Spock extension to install. |
+| pg_port | 5432 | Port where Postgres should listen for connections. |
+| pg_path | `/usr/pgsql-${pg_version}` on RHEL, `/usr/lib/postgresql/${pg_version} on Debian | Where Postgres binary and library directories can be found. Almost all systems are based on RHEL or Debian variants. Specify for custom location. |
```
You can ignore this for now, the documentation is being fully rewritten in another PR, including the README, which removes the entire section you corrected.

paul-pgedge commented:
I did some initial testing on Rocky 9. Will be testing with Ubuntu 24.04 next to verify Debian steps. Here's what I've found

### `pg_data`

The pg_data setting doesn't seem to work. I specified it as `/data/pgedge` and changed the ownership and permissions but I got this error when I ran it the first time:

```
An exception occurred during task execution. To see the full traceback, use -vvv. The error was: OSError: [Errno 18] Invalid cross-device link: b'/tmp/.ansible_tmp4gr607ee' -> b'/data/pgedge//server.key'

fatal: [node0]: FAILED! => {"changed": false, "msg": "The destination directory (/data/pgedge) is not writable by the current user. Error was: [Errno 13] Permission denied: b'/data/pgedge/.ansible_tmpjp6m743zserver.key'"}

An exception occurred during task execution. To see the full traceback, use -vvv. The error was: OSError: [Errno 18] Invalid cross-device link: b'/tmp/.ansible_tmpvniz9ii7' -> b'/data/pgedge//server.key'

fatal: [node1]: FAILED! => {"changed": false, "msg": "The destination directory (/data/pgedge) is not writable by the current user. Error was: [Errno 13] Permission denied: b'/data/pgedge/.ansible_tmp62k1r29eserver.key'"}
```

So I set the permissions on that directory to 777 and retried it and got this error:

```
TASK [setup_postgres : Configure Postgres for pgEdge compatibility] ***********************************************************************************************************
fatal: [node0]: FAILED! => {"changed": false, "msg": "Path /data/pgedge//postgresql.conf does not exist !", "rc": 257}
fatal: [node1]: FAILED! => {"changed": false, "msg": "Path /data/pgedge//postgresql.conf does not exist !", "rc": 257}
```

But when I check the default location - `/var/lib/pgsql/17/data` - everything is there and not in the configured directory.

The behavior is the same if I set the `pg_data` at the role level.

### haproxy_routes

I haven't tested this yet because I'm not sure how to format them. Additional examples would be good to add.

### custom_hba_conf

It took me a bit of digging through the Ansible docs to figure out how to use `custom_hba_conf`, another place some examples would help.

### README updates

There are a few places in the README that refer to `$cluster_path` that should be changed since we're not using the cluster anymore.

----

### Inventory Used:

```yaml
Inventory Used:

```yaml
# This inventory will produce an HA cluster with the following:
#
# * 2 zones
# * 3 pgEdge nodes in each zone
# * 1 HAProxy node in each zone

all:
  vars:
    pg_version: 17
    debug_pgedge: false
    repo_name: release
    db_user: admin
    db_password: Really_s3cure_password
    spock_version: 5.0.4
    is_ha_cluster: true
    pg_data: /data/pgedge/
    custom_hba_rules:
      - users: admin
        databases: mattermost,keycloak,pgbench,postgres
        source: 10.0.0.0/24
    db_names:
      - mattermost
      - keycloak
      - pgbench

pgedge:
  vars:
    ansible_user: rocky
    pg_data: /data/pgedge/
  hosts:
    node0:
      zone: 2
    node1:
      zone: 1
    node2:
      zone: 2
    node3:
      zone: 1
    node4:
      zone: 2
    node5:
      zone: 1


haproxy:
  vars:
    ansible_user: rocky
  hosts:
    node0:
      zone: 2
    node1:
      zone: 1
    node2:
      zone: 2
    node3:
      zone: 1
    node4:
      zone: 2
    node5:
      zone: 1


backup:
  vars:
    ansible_user: rocky
  hosts:
    backrest0:
      zone: 1
    backrest1:
      zone: 2
```

bonesmoses committed:
Now properly bootstraps pg_data path and directs service to use that location

This required writing separate Debian and RHEL Postgres init steps due to how
those systems work.

bonesmoses committed:
Fixed issue where pg_createcluster was called with invalid parameters.

bonesmoses commented on roles/install_repos/tasks/rhel_epel_dl.yaml:
```diff
@@ -0,0 +1,15 @@
+
+# Most RHEL variants need to install the EPEL repo from a remote fetch.
+
+- name: Install EPEL repo package for enterprise Linux systems
+  dnf:
+    name: "https://dl.fedoraproject.org/pub/epel/epel-release-latest-{{ os_version }}.noarch.rpm"
+    state: present
+  vars:
+    os_version: "{{ ansible_facts.distribution_major_version }}"
+  retries: 5
+  delay: 20
+  register: result
+  until: result is success
+  become: true
```
This file is only included by `rhel_epel.yaml` if the major version is > 7, so this can't actually happen.

bonesmoses commented on roles/setup_postgres/tasks/prep_instance.yaml:
```diff
@@ -0,0 +1,113 @@
+  become: true
+  become_user: postgres
+
+- name: Add access for the pgedge communication and admin/DB user
+  community.postgresql.postgresql_pg_hba:
+    dest: "{{ pg_data }}/pg_hba.conf"
+    contype: host
+    users: "{{ pgedge_user + ',' + db_user }}"
+    databases: "{{ (db_names + ['postgres']) | join(',') }}"
+    method: scram-sha-256
+    source: "{{ hostvars[item].ansible_default_ipv4.address }}/32"
+  loop: "{{ groups['pgedge'] }}"
+  become: true
+  become_user: postgres
```
It's not just this task; there are several implicit uses of ipv4 through the collection. I think we'll leave this as is for now.

bonesmoses committed:
Some fixes based on Coderabbit feedback

* Bumped community.crypto version requirement.
* Fixed empty default in init_server role.
* Added missing type conversion to distribution version checks.
* Switched to simpler ansible_* variables from ansible_facts where possible.
* No longer assume x86_64 during EPEL installation on RHEL 10 systems.
* Revised some potentially misleading code comments.

vyruss commented on roles/init_server/tasks/main.yaml:
```diff
@@ -1,13 +1,28 @@
+# Only pgEdge and backup servers should need an SSH key for communication.
+# This step is done here so all keys exist for all servers before we start
+# adding them to known hosts and authorized key files.
+
```
If a host is both in `pgedge` & `backup` groups we end up overwriting the key. Besides, this is bad practice we recommend against.
```
# Backup servers should be dedicated only for backup use (not DB servers)
- name: Ensure no host is in both the pgedge and backup groups
  assert:
    that:
      - groups.pgedge | default([]) | intersect(groups.backup | default([])) | length == 0
    fail_msg: "A host cannot be in both the 'pgedge' and 'backup' groups."
  run_once: true
```

vyruss commented on roles/setup_backrest/tasks/setup_server.yaml:
```diff
@@ -1,27 +1,55 @@
+
+# With configuration complete, attempt to take the initial bootstrap backup:
+
+- include_tasks: first_backup.yaml
```
let's pass this down to `first_backup.yaml`
```
  vars:
    os_user: "{{ backup_repo_user }}"
```

vyruss commented on roles/setup_backrest/tasks/setup_client.yaml:
```diff
@@ -0,0 +1,54 @@
+# bootstrap backup.
+
+- include_tasks: first_backup.yaml
+  when:
```
let's pass this down to `first_backup.yaml`
```
  vars:
    os_user: "{{ backup_repo_user }}"
```

vyruss commented on roles/setup_backrest/tasks/first_backup.yaml:
let's make sure we are the correct backup user
```
  become: true
  become_user: "{{ os_user }}"
```

vyruss commented on roles/setup_backrest/tasks/first_backup.yaml:
let's make sure we are the correct backup user
```
  become: true
  become_user: "{{ os_user }}"
```

vyruss commented on roles/install_pgedge/tasks/main.yaml:
```diff
@@ -1,11 +1,15 @@
+    name:
+    - "pgedge-enterprise-all{{ separator }}{{ pg_version }}"
+    - pgedge-python3-psycopg2
+    state: latest
```
is it intentional to pull in the latest `pgedge-enterprise-all-` package with every run? (`state: latest`)

vyruss commented on roles/install_etcd/tasks/install.yaml:
These should be `root:root` for security (i.e. etcd pwnage)
```
    owner: root
    group: root
    mode: "0644"
```

vyruss commented on roles/init_server/tasks/main.yaml:
```diff
@@ -1,13 +1,28 @@
 ---
 
```
Let's not allow creating a cluster with default passwords.

```
# Disallow creating a cluster with default passwords.
- name: Ensure default passwords have been changed
  assert:
    that:
      - db_password != 'secret'
      - pgedge_password != 'secret'
      - replication_password != 'secret'
      - backup_password != 'secret'
    fail_msg: >-
      One or more passwords are still set to the default value 'secret'. Please set secure passwords.
  run_once: true
```

vyruss commented on roles/init_server/tasks/platform.yaml:
```diff
@@ -0,0 +1,20 @@
+---
+
+# The RemoveIPC setting is probably the most idiotic systemd feature to date.
```
suggested rewording:
```
# The RemoveIPC setting in systemd is dangerous for databases.
# When enabled (which is the DEFAULT on Debian systems), logging out of a shell
# will trigger systemd to deallocate all shared memory segments owned by that user.
# This can instantly crash Postgres, so we must ensure this is disabled globally.
```

vyruss commented on roles/setup_pgedge/vars/main.yaml:
```diff
@@ -1,15 +1,8 @@
-    map('extract', hostvars) |
-    selectattr('zone', 'eq', zone) |
-    map(attribute='inventory_hostname') | list }}
+    {{ hostvars.values() | map(attribute='zone', default='') |
```
this can cause the `zone_list` to contain empty string, do we want that?

vyruss replied:
ok disregard

vyruss commented on roles/setup_postgres/tasks/replica_setup.yaml:
```diff
@@ -0,0 +1,27 @@
+    state: absent
+  become: true
+  when:
+  - ansible_facts.services['patroni.service']['state'] | default('') != 'running'
```
This can blow away the data dir just because patroni isn't running? Or am I misreading it?

vyruss replied:
Scenario: node is replicating fine, but patroni has somehow died. In an attempt to restore patroni, user re-runs ansible and destroys a perfectly fine PGDATA

bonesmoses replied:
I've just removed the directory cleanup in my draft logic. That said, while revising the role to use the PGPDG packaged Patroni, these steps will require an overhaul anyway.

vyruss commented on galaxy.template.yml:
url fix
```
documentation: https://github.com/pgEdge/pgedge-ansible/blob/main/README.md
```

vyruss reviewed (commented):
some discussion points raised

bonesmoses committed:
To address code review:

* Bumped community.crypto version requirement.
* Fixed empty default in init_server role.
* Added missing type conversion to distribution version checks.
* Switched to simpler ansible_* variables from ansible_facts where possible.
* No longer assume x86_64 during EPEL installation on RHEL 10 systems.
* Revised some potentially misleading code comments.
* Switch to Patroni from PGDG repository.
* Fixed a bug that was causing Ansible to SELinux disable step when requested.

bonesmoses committed:
* Moved patroni_bin_dir to vars/ so it's role managed rather than a default.

dpage commented:
Claude review:

  High-Level Architecture Feedback
                                                                                                                                                                                                                                                                                                                       
  The direction is sound - moving to native packages and standard Postgres conventions is the right call for production use. However, the sheer scope (102 files) makes this very hard to review confidently. Consider whether it could be split into smaller, sequentially-mergeable PRs (e.g., repo setup, package
  install, TLS, Patroni refactor, Spock refactor).                                                                                                                                                                                                                                                                     
                  
  ---                                                                                                                                                                                                                                                                                                                  
  Bugs / Correctness Issues
                                                                                                                                                                                                                                                                                                                       
  1. Inconsistent Snowflake parameter name - roles/setup_postgres/tasks/prep_instance.yaml:46 sets snowflake.zone = {{ zone }} but the Patroni template (roles/setup_patroni/templates/patroni.yml.j2:66) uses snowflake.node: {{ zone }}. One of these is wrong and will cause the snowflake extension to not
  configure correctly on either HA or non-HA clusters.                                                                                                                                                                                                                                                                 
  2. Undefined variable fallthrough on unsupported OS families - roles/role_config/defaults/main.yaml uses conditional Jinja for pg_home, pg_path, and pg_data that only handles RedHat and Debian. If ansible_os_family is anything else (e.g., Suse), these resolve to False/empty string, causing silent downstream
  failures. The new validate.yaml doesn't check for supported OS families.                                                                                                                                                                                                                                             
  3. LD_LIBRARY_PATH may be unnecessary - roles/setup_backrest/tasks/first_backup.yaml still sets LD_LIBRARY_PATH: "{{ pg_path }}/lib" in the environment, but since pgbackrest is now installed as a system package (pgedge-pgbackrest), it should already link to the correct libraries. If it is still needed, this
  is worth a comment explaining why.                                                                                                                                                                                                                                                                                   
  4. Password validation may fail when backup group is absent - roles/init_server/tasks/validate.yaml:40 checks backup_password != 'secret' on all hosts in groups.pgedge | union(groups.backup | default([])), but backup_password is defined in role_config defaults so this should be fine. However, the when
  condition runs on backup hosts too, which may not have all variables if they don't include role_config.                                                                                                                                                                                                              
  5. shared_preload_libraries order changed - The Patroni template changed the order from "pg_stat_statements, snowflake, spock" to "spock, snowflake, pg_stat_statements". The non-HA prep_instance.yaml also uses the new order. If Spock has load-order dependencies, this could matter. Is this intentional?
  6. Merge conflicts - The PR is in a dirty mergeable state and needs to be rebased/merged with current main.                                                                                                                                                                                                          
                                                                                                                                                                                                                                                                                                                       
  ---                                                                                                                                                                                                                                                                                                                  
  Security Concerns                                                                                                                                                                                                                                                                                                    
                   
  7. GPG check disabled for Rocky 9 PGDG repo - roles/install_repos/tasks/rhel_repo.yaml:27-29 disables GPG validation for the PGDG repo on Rocky 9. Even with the explanatory comment, this is a supply-chain risk. Consider pinning the specific RPM hash or documenting a workaround timeline.
  8. TLS artifacts in working directory - tls/etcd/ and tls/postgres/ directories are created on the Ansible controller in the playbook working directory. These contain private keys and could easily be committed to git. Add tls/ to .gitignore (alongside the existing host-keys/ pattern if applicable).          
  9. backup_repo_user defaults to ansible_user_id - roles/role_config/defaults/main.yaml:57 sets backup_repo_user: "{{ ansible_user_id }}". If someone runs the playbook as root, this creates a root-owned backup repo, which is probably not intended.                                                               
                                                                                                                                                                                                                                                                                                                       
  ---                                                                                                                                                                                                                                                                                                                  
  Idempotency / Operational Issues                                                                                                                                                                                                                                                                                     
                                                                                                                                                                                                                                                                                                                       
  10. systemd-logind restart on every run - roles/init_server/tasks/platform.yaml:15-18 unconditionally restarts systemd-logind even if RemoveIPC=no was already set. This can kill active user sessions. Use a register + when: result.changed pattern, or use a handler.
  11. state: latest on prereq packages - roles/install_repos/tasks/debian_repo.yaml:5 uses state: latest for curl/gnupg2/lsb-release, which always shows "changed" and triggers unnecessary work. Use state: present for idempotency unless you explicitly need the latest version.                                    
  12. HAProxy only restarts on config change, but first run may not start - roles/setup_haproxy/tasks/setup.yaml uses state: "{{ 'restarted' if config_status.changed else 'started' }}". This is good, but on first run config_status.changed is true so it restarts a service that may not yet be started. This      
  should be fine with restarted but worth verifying.                                                                                                                                                                                                                                                                   
                                                                                                                                                                                                                                                                                                                       
  ---                                                                                                                                                                                                                                                                                                                  
  Style / Minor Issues
                      
  13. roles/install_repos/tasks/debian_repo.yaml:53 uses command with creates for the PGDG repo setup, which is good for idempotency.
  14. Template file locations - Good cleanup moving from files/*.j2 to templates/*.j2, which follows Ansible conventions.                                                                                                                                                                                              
  15. New dependencies - community.general and community.crypto are added to galaxy.template.yml. Make sure the minimum versions specified (>=8.2.0 and >=2.4.0) actually include the modules used (e.g., dnf_config_manager, TLS modules).                                                                            
  16. README.md - There's a double blank line added at line 117 in the Usage section. Minor, but messy.                                                                                                                                                                                                                
  17. pgbackrest.conf.j2 - The pg1-user=postgres is hardcoded rather than using {{ backup_user }} or a variable. This may be intentional (since the postgres user owns the data directory), but it's worth clarifying in a comment.                                                                                    
                                                                                                                                                                                                                                                                                                                       
  ---                                                                                                                                                                                                                                                                                                                  
  Questions for the Author                                                                                                                                                                                                                                                                                             
                          
  - Is the snowflake.zone vs snowflake.node discrepancy a bug or was the parameter renamed in a newer Spock version?
  - What's the plan for the tls/ directory on the controller? Should it be gitignored, or is there a cleanup step?                                                                                                                                                                                                     
  - Has this been tested on both RHEL and Debian with both HA and non-HA configurations end-to-end?                                                                                                                                                                                                                    
  - The PR has been open since November 2024. Is the conflict resolution straightforward, or has main diverged significantly?

bonesmoses committed:
* Removed unnecessary LD_LIBRARY_PATH calls
* Now using proper snowflake.node argument in non-HA setup
* Git ignore tls and host-keys directories which may contain generated assets
* Exeplictly require RedHat or Debian Linux distribution families
* No longer require "latest" repo packages to avoid forced upgrades
* Only restart logind if we changed the IPC state

bonesmoses committed:
Merge branch 'main' into feature/EE-22/switch-to-system-packages

bonesmoses committed:
* Use correct default cluster path when using non-default name on Debian.
* Use correct pg_hba.conf location on Debian systems.
* Fixed an issue against new versions of ansible that when requires booleans.
* Removed uncaught legacy use of patroni_bin_dir.
* Replaced old patroni status check with intended include version.

bonesmoses committed:
* Use correct pg_hba.conf location on Debian systems.

bonesmoses committed:
* Fixed hard-coded patroni cluster context name.

vyruss reviewed (approved):
* Feedback provided and all code issues have been addressed.
* End-to-end tests for sample scenarios pass on Debian 12 and Rocky 9.

✅ from me
